torch/csrc/fx/node.cpp

```diff
@@ -0,0 +1,241 @@
+#include <torch/csrc/fx/node.h>
+
+#include <structmember.h>
+
+////////////////////////////////
+// NodeBase
+///////////////////////////////
+
+struct NodeBase {
+  PyObject_HEAD bool _erased;
+  NodeBase* _prev;
+  NodeBase* _next;
+};
+
+static PyObject* NodeBase_new(
+    PyTypeObject* type,
+    PyObject* args,
+    PyObject* kwds) {
+  PyObject* self = type->tp_alloc(type, 0);
+  if (!self)
+    return nullptr;
+  return self;
+}
+
+static int NodeBase_init_fn(NodeBase* self, PyObject* args, PyObject* kwds) {
+  self->_erased = false;
+  Py_INCREF(self);
+  self->_prev = self;
+  Py_INCREF(self);
+  self->_next = self;
+  return 0;
+}
+
+static void NodeBase_dealloc(NodeBase* self) {
+  Py_XDECREF(self->_prev);
+  Py_XDECREF(self->_next);
+  Py_TYPE(self)->tp_free((PyObject*)self);
+}
```

Comment:
There appears to be a memory leak, I have done some ablations in order to figure out the cause, it shows that it is coming from NodeBase, and most specifically from `_prev` and `_next`. I suspect them being initialiazed to self is not jiving well with the gc somehow but not really sure how to prove. If anyone has an idea on how to figure this out, i'd appreciate it.

Reply:
Well, I mean, if you initialize the field to self then you have a reference cycle, and it won't get eliminated until you GC, which could trigger a weakref test

Reply:
This was my first time writing something with c api, did you mean someone how marking these pointers as weak ref? any docs on this? Googling "python c api weak ref" did not give me much

Reply:
Oh but I guess your point is the old code also generated cycles and was not a problem

Reply:
Check https://github.com/pytorch/pytorch/pull/128288/files#r1632406879

Reply:
You could refer to the implementation of `_functools.lru_list_item` and `_functools.lru_cache_object` for reference.

https://github.com/python/cpython/blob/35c799d79177b962ddace2fa068101465570a29a/Modules/_functoolsmodule.c#L761-L776